Text fetching

Fantaskink opened the pull request:
This pull request introduces a new `Fetcher` class for retrieving and processing book text from the Gutendex API, along with comprehensive tests for its functionality. It also adds the `requests` library as a dependency and updates the project structure to support these changes.

**Text fetching and processing functionality:**

* Added `Fetcher` class in `src/text_fetching/fetcher.py` that can fetch a random book's text from the Gutendex API, extract a random slice from the text, and format text to only lowercase alphabetic characters.

**Testing:**

* Added a suite of tests in `test/text-fetching/test_fetcher.py` to validate the `Fetcher` class methods, including edge cases for invalid input and text formatting.

**Dependency management:**

* Updated `pyproject.toml` to include `requests` as a required dependency for making HTTP requests.

**Project structure:**

* Modified `src/main.py` to remove the placeholder print statement, preparing it for future logic integration.

Fantaskink committed:
Add book slice fetcher class

Fantaskink committed:
Merge remote-tracking branch 'origin/dev' into text-fetching

Fantaskink committed:
Add type annotations

Fantaskink committed:
Move fetcher to src, add dependencies

Fantaskink committed:
Use lower instead of uppercase for plaintext

Fantaskink committed:
Set up test directory

Fantaskink committed:
Remove unused code

Fantaskink committed:
Add tests for fetcher.py

Fantaskink committed:
Merge branch 'dev' into text-fetching

Fantaskink committed:
Add requests dependency

Fantaskink committed:
Remove unused imports

Jatewo commented on src/main.py:
Perhaps cut down on the print use.

Jatewo commented on src/main.py:
```diff
@@ -1,6 +1,22 @@
+
 def main():
-    print("Hello from ciphertext-generation!")
+	print("Hello from ciphertext-generation!")
```
Remove?

Jatewo commented on src/main.py:
```diff
@@ -1,6 +1,22 @@
+	book_text = fetcher.fetch_random_book_text()
+	if book_text:
+		print("Fetched book text successfully!")
+		print(f"First 500 characters:\n{book_text[:500]}")
```
Remove?

Jatewo commented on src/main.py:
```diff
@@ -1,6 +1,22 @@
+			print(f"\nRandom slice ({len(slice_text)} characters):\n{slice_text}")
+			formatted = fetcher.format_text(slice_text)
+			print(f"\nFormatted text length (alphabetic chars only): {len(formatted)}")
+			print(f"First 100 characters of formatted text: {''.join(formatted[:100])}")
```
Remove?

Jatewo commented on src/main.py:
```diff
@@ -1,6 +1,22 @@
+		print(f"First 500 characters:\n{book_text[:500]}")
+		slice_text = fetcher.get_random_book_slice(book_text)
+		if slice_text:
+			print(f"\nRandom slice ({len(slice_text)} characters):\n{slice_text}")
```
Remove?

Jatewo commented on src/text_fetching/fetcher.py:
I assume all the prints are for debugging. Perhaps remove?

Jatewo commented on src/text_fetching/fetcher.py:
```diff
@@ -0,0 +1,96 @@
+			return book_text
+			
+		except requests.RequestException as e:
+			print(f"Error fetching book data: {e}")
```
Perhaps raise an error here instead

Fantaskink committed:
Run formatter on 3 files

Jatewo commented on src/text_fetching/fetcher.py:
```diff
@@ -0,0 +1,96 @@
+		
+	def format_text(self, text: str) -> str:
+		if not isinstance(text, str):
+			raise ValueError(
```
Unreachable code

Fantaskink replied:
To my understanding, this is not unreachable, since python allows you to call the function with any type.

Jatewo commented on src/text_fetching/fetcher.py:
```diff
@@ -0,0 +1,96 @@
+			return None
+
+	def get_random_book_slice(self, book_text: str, min_len: int = 100, max_len: int = 5000) -> str | ValueError:
+		"""Extract a random slice from the provided book text."""
```
Consider autoDocstring extension to help generate these descriptions (adds descriptions for arguments and return type as well)

Fantaskink replied:
Done

Mafusn commented on test/text-fetching/test_fetcher.py:
```diff
@@ -0,0 +1,78 @@
+    with pytest.raises(ValueError) as excinfo:
+        fetcher.format_text(no_text)
+    assert (
+        "text must a string"
```
"text must a string" 🍄

Fantaskink replied:
Fixed

Jatewo commented on test/text-fetching/test_fetcher.py:
Add tests for remaining function

Mafusn reviewed (approved):
LGTM 🦅

Fantaskink committed:
Remove excessive print statements

Fantaskink committed:
Fix typo

Fantaskink committed:
Remove unused import

Fantaskink committed:
Add AutoDocString documentation

Fantaskink committed:
Fix typo

Jatewo commented on src/text_fetching/fetcher.py:
```diff
@@ -0,0 +1,111 @@
+            return book_text
+
+        except requests.RequestException as e:
+            print(f"Error fetching book data: {e}")
```
Raise error

Jatewo commented on test/text-fetching/test_fetcher.py:
```diff
@@ -0,0 +1,78 @@
+        book_text=long_text, min_len=100, max_len=5000)
+    assert len(sliced_text) <= 5000
+    assert len(sliced_text) >= 100
+    assert isinstance(sliced_text, str)
```
This line should be above len or it might fail on len

Fantaskink committed:
Raise error instead of returning None

Fantaskink committed:
Move line in test

Jatewo committed:
test: Add pytest-mock and expand fetcher test coverage

Added pytest-mock to dev dependencies in pyproject.toml. Extended test_fetcher.py with tests for Fetcher.fetch_random_book_text, including success, missing format, and request exception scenarios using pytest-mock.

Jatewo reviewed (approved):
LGTM dawg 👍

Mafusn reviewed (approved):
LGTM 🌮
